packages/tldraw/src/lib/shapes/shared/useAsset.ts

```diff
@@ -19,81 +25,107 @@ import { useEffect, useRef, useState } from 'react'
  *
  * @public
  */
-export function useAsset(options: {
-	shapeId: TLShapeId
-	assetId: TLAssetId | null
-	width: number
-}) {
-	const { shapeId, assetId, width } = options
+export function useAsset({ shapeId, assetId }: { shapeId: TLShapeId; assetId: TLAssetId | null }) {
 	const editor = useEditor()
-	const [url, setUrl] = useState<string | null>(null)
-	const [isPlaceholder, setIsPlaceholder] = useState(false)
 	const isExport = !!useSvgExportContext()
-	const asset = assetId ? editor.getAsset(assetId) : null
-	const culledShapes = editor.getCulledShapes()
-	const isCulled = culledShapes.has(shapeId)
-	const didAlreadyResolve = useRef(false)
 	const isReady = useDelaySvgExport()
 
-	useEffect(() => {
-		if (url) didAlreadyResolve.current = true
-	}, [url])
+	// We use a state to store the result of the asset resolution, and we're going to avoid updating this whenever we can
+	const [result, setResult] = useState<{
+		asset: (TLImageAsset | TLVideoAsset) | null
+		url: string | null
+	}>(() => ({
+		asset: assetId ? editor.getAsset<TLImageAsset | TLVideoAsset>(assetId) ?? null : null,
+		url: null as string | null,
+	}))
 
-	const shapeScale = asset && 'w' in asset.props ? width / asset.props.w : 1
-	// We debounce the zoom level to reduce the number of times we fetch a new image and,
-	// more importantly, to not cause zooming in and out to feel janky.
-	const screenScale = useValue('zoom level', () => editor.getZoomLevel() * shapeScale, [
-		editor,
-		shapeScale,
-	])
+	// A flag for whether we've resolved the asset URL at least once, after which we can debounce
+	const didAlreadyResolve = useRef(false)
 
-	useEffect(() => {
-		if (url) didAlreadyResolve.current = true
-	}, [url])
+	// The last URL that we've seen for the shape
+	const previousUrl = useRef<string | null>(null)
 
 	useEffect(() => {
-		if (!isExport && isCulled) return
+		if (!assetId) return
 
-		if (assetId && !asset?.props.src) {
-			const preview = editor.getTemporaryAssetPreview(assetId)
+		let isCancelled = false
+		let cancelDebounceFn: (() => void) | undefined
 
-			if (preview) {
-				setUrl(preview)
-				setIsPlaceholder(true)
-				isReady()
-				return
-			}
-		}
+		const cleanupEffectScheduler = react('update state', () => {
```

Comment:
your previous work on this, in https://github.com/tldraw/tldraw/pull/4045, you had it using `useReactor` - do you think that's a better tact or is this take more what you were looking for? wdyt?

Reply:
I think we might be able to use that. Originally there was some extra logic before the `react` that made me use the effect instead, but now it's pretty cleaned up